Clang-8 incompatabilities

After adding all the backend changes to be able to use clang-8 in the CI the following problems appear.

Linux: https://oe-jenkins-tf.westeurope.cloudapp.azure.com/job/pipelines/job/Azure-Linux/252/

` 75/280 Test  #75: tests/cppException ....................................................................................................................Child aborted***Exception:   0.18 sec
=== This program is used to test basic cpp exception functionalities.
=== TestCppException() 
char type exception tests passed.
integer type exception tests passed.
string type exception tests passed.
string reference type exception tests passed.
user defined class type exception tests passed.
user defined class reference type exception tests passed.
user defined class type exception test passed.
user defined class reference type exception test passed.
user defined class type exception test passed.
Ellipsis exception catch tests passed.
Re-throw and throw new exception in the catch clause tests passed.
catch clause in the outermost block tests passed.
Find the matching catch clause through call stack tests passed
StackUnwind local unwind test passed.
StackUnwind global unwind test passed.
StackUnwind local unwind second test passed.
StackUnwind global unwind second test passed.
StackUnwind unwind test passed.
FunctionTryBlock exception test passed.
Cpp exception tests passed!
=== passed regular cpp exception tests.
=== TestUnhandledException(0)  
This test will crash the enclave.
2020-09-29T20:12:25+0000.765056Z [(H)ERROR] tid(0x7f8e722ab740) | :OE_ENCLAVE_ABORTING [../host/calls.c:_call_enclave_function_impl:56]
=== passed unhandled cpp exception tests (0).
=== TestUnhandledException(1)  
This test will crash the enclave.
        Start  76: tests/crypto_crls_cert_chains
`

This seems to be only when the build configuration is 'RelWithDebInfo'.

However, Windows fails to build entirely in `Release` mode

Windows: https://oe-jenkins-tf.westeurope.cloudapp.azure.com/job/pipelines/job/Azure-Windows/274/

`[2809/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\search\tsearch.c.obj
[2810/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\search\twalk.c.obj
FAILED: libc/CMakeFiles/oelibc.dir/__/3rdparty/musl/musl/src/search/twalk.c.obj 
"C:/Program Files/Git/bin/bash.exe" "C:/Users/oeadmin/workspace/pipelines/Azure-Windows/scripts/clangw" "-DOE_API_VERSION=2 -DOE_BUILD_ENCLAVE -D_XOPEN_SOURCE=700 -I3rdparty\musl\musl\src\include -I3rdparty\musl\musl\src\internal -I3rdparty\musl\musl\arch\x86_64 -I..\..\include -Ioutput\include -Ioutput\include\openenclave\libc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 -Wall -Werror -Wpointer-arith -Wconversion -Wextra -Wno-missing-field-initializers -fno-strict-aliasing -mxsave -fno-builtin-malloc -fno-builtin-calloc -fno-builtin -mllvm -x86-speculative-load-hardening -fno-builtin-memcpy -fno-builtin-memset -ffreestanding -mrdrnd -Wno-missing-braces -Wno-parentheses -Wno-unknown-pragmas -Wno-conversion -Wno-unused-parameter -Wno-sign-compare -Wno-dangling-else -Wno-ignored-attributes -Wno-logical-op-parentheses -Wno-string-plus-int -Wno-typedef-redefinition -Wno-unneeded-internal-declaration -m64 -fPIE -nostdinc -fstack-protector-strong -fvisibility=hidden -fno-omit-frame-pointer -ffunction-sections -fdata-sections -ftls-model=local-exec /showIncludes -o libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\search\twalk.c.obj -c ..\..\3rdparty\musl\musl\src\search\twalk.c"
fatal error: error in backend: Undefined temporary symbol
[2811/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\stat\fstat.c.obj
[2812/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\stdio\asprintf.c.obj
[2813/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\stdio\dprintf.c.obj
[2814/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\stdio\clearerr.c.obj
[2815/4286] Building C object libc\CMakeFiles\oelibc.dir\__\3rdparty\musl\musl\src\stdio\fclose.c.obj
ninja: build stopped: subcommand failed.
script returned exit code 1`

Additionally Code Coverage is only compatible with clang-7 at this time.

--- a/3rdparty/openssl/CMakeLists.txt
+++ b/3rdparty/openssl/CMakeLists.txt
@@ -78,10 +78,12 @@ ExternalProject_Add(
   COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD_INCLUDES}/openssl
   COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/bn_conf.h
           ${OPENSSL_BUILD_INCLUDES}/crypto/bn_conf.h
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/dso_conf.h
-          ${OPENSSL_BUILD_INCLUDES}/crypto/dso_conf.h
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/opensslconf.h
-          ${OPENSSL_BUILD_INCLUDES}/openssl/opensslconf.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/dso_conf.h
+    ${OPENSSL_BUILD_INCLUDES}/crypto/dso_conf.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/opensslconf.h
+    ${OPENSSL_BUILD_INCLUDES}/openssl/opensslconf.h
   BUILD_COMMAND ""
   INSTALL_COMMAND "")
 
@@ -101,14 +103,21 @@ ExternalProject_Add(
     ${CMAKE_COMMAND} -E copy ${OPENSSL_BUILD_INCLUDES}/openssl/opensslconf.h
     ${OPENSSL_INSTALL_INCLUDES}/opensslconf.h
   COMMAND ${OE_BASH} -c "${OPENSSL_APPEND_UNSUPPORTED}"
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/unsupported.h
-          ${OPENSSL_INSTALL_INCLUDES}/unsupported.h
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/crypto_unsupported.h
-          ${OPENSSL_INSTALL_INCLUDES}/crypto_unsupported.h
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/ssl_unsupported.h
-          ${OPENSSL_INSTALL_INCLUDES}/ssl_unsupported.h
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/x509_vfy_unsupported.h
-          ${OPENSSL_INSTALL_INCLUDES}/x509_vfy_unsupported.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/include/unsupported.h
+    ${OPENSSL_INSTALL_INCLUDES}/unsupported.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy
+    ${CMAKE_CURRENT_SOURCE_DIR}/include/crypto_unsupported.h
+    ${OPENSSL_INSTALL_INCLUDES}/crypto_unsupported.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy
+    ${CMAKE_CURRENT_SOURCE_DIR}/include/ssl_unsupported.h
+    ${OPENSSL_INSTALL_INCLUDES}/ssl_unsupported.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy
+    ${CMAKE_CURRENT_SOURCE_DIR}/include/x509_vfy_unsupported.h
+    ${OPENSSL_INSTALL_INCLUDES}/x509_vfy_unsupported.h
   BUILD_COMMAND ""
   INSTALL_COMMAND "")
 
@@ -943,8 +952,11 @@ enclave_compile_options(opensslssl PRIVATE ${OPENSSL_COMPILE_OPTIONS})
 
 # Clang verion greater than 8 fails on compiling the following soruce with optimization.
 # Explicitly set -O0 as a workaround. See #2556 for the similar issue.
-if (OPENSSL_C_COMPILER MATCHES clang AND CMAKE_C_COMPILER_VERSION
-                                         VERSION_GREATER_EQUAL 8)
+# Always dose this on Windows as we cannot know the version of clang.
+if (WIN32
+    OR (OPENSSL_C_COMPILER MATCHES clang
+        AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 8
+        AND CMAKE_C_COMPILER_VERSION VERSION_LESS 9.0))
   set_property(
     SOURCE openssl/crypto/modes/cts128.c
     APPEND_STRING

--- a/libc/CMakeLists.txt
+++ b/libc/CMakeLists.txt
@@ -867,9 +867,11 @@ elseif (CMAKE_C_COMPILER_ID MATCHES Clang OR USE_CLANGW)
 endif ()
 
 # Disable optimizations for twalk.c to avoid Clang 8 compiler bug, see #2556
-if (CMAKE_C_COMPILER_ID MATCHES Clang
-    AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 8
-    AND CMAKE_C_COMPILER_VERSION VERSION_LESS 9.0)
+# Always dose this on Windows as we cannot know the version of clang.
+if (WIN32
+    OR (CMAKE_C_COMPILER_ID MATCHES Clang
+        AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 8
+        AND CMAKE_C_COMPILER_VERSION VERSION_LESS 9.0))
   set_source_files_properties(${MUSLSRC}/search/twalk.c PROPERTIES COMPILE_FLAGS
                                                                    -O0)
 endif ()

